### PR TITLE
Adding jinc function

### LINKED ISSUE #264: Add jinc function
Hey,

I needed a jinc function $\text{jinc} = \frac{J_1(x)}{x}$ which is not registered.
I would create a pull request, but we maybe we can discuss how it should look like. My current code is
```julia
using SpecialFunctions

function jinc(x::T) where T
    if iszero(x)
        return convert(T, 0.5)
    else
        return besselj1(x) / x 
    end 
end
```

I do not really like this convert and it would throw an error for `jinc(0)` but for me that's good.
I've seen you're using rather [this](https://github.com/JuliaMath/SpecialFunctions.jl/blob/9116108ba52d103a05d47d9af1ca6b959a536c30/src/bessel.jl#L185), so should I create something similar for `jinc`?

Cheers,

Felix

### PR DESCRIPTION
**Pull request by roflmaostc**
Hey,

we already discussed a bit in #264 about adding a `jinc` function.
We decided on the following convention, which is similar to `sinc` in terms of the peak height and argument scaling: 

&space;=&space;\frac{2\cdot&space;J_1(\pi&space;x)}{\pi&space;x})

I wanted to use `fastabs` which is only available in the current master of julia. This is why I copied the definition to the if statement.

I would be happy to receive some feedback!

Thanks,

Felix

(Closes #264.)

### CODE REVIEW

**Comment by roflmaostc** [comment]
Error for 1.3: `UndefVarError: evalpoly not defined`

`evalpoly` needs `julia >= 1.4`. What should we do about that?

File: src/bessel.jl [hunk 1]
```diff
@@ -758,3 +758,32 @@ External links: [DLMF](https://dlmf.nist.gov/10.2.6), [Wikipedia](https://en.wik
...
+ # for small arguments, a Taylor series is faster
+@inline _jinc(x::Union{T,Complex{T}}) where {T<:Union{Float32,Float64}} = begin
+    # replace in future with fastabs
+    # https://github.com/JuliaLang/julia/blob/93fb785831dcfcc442f82fab8746f0244c5274ae/base/special/trig.jl#L1057
+    if abs(real(x)) + abs(imag(x)) < _jinc_threshold(T)
+        return evalpoly(x^2, (T(1), -T(π)^2/8, T(π)^4/192))
+    else
+        return _jinc_core(x)
+    end
+end
```
Comment (giordano) [inline src/bessel.jl:787]: This is a fun way to define a function.  Note that the `begin`/`end` block is not needed at all
```julia
@inline _jinc(x::Union{T,Complex{T}}) where {T<:Union{Float32,Float64}} = 
    if abs(real(x)) + abs(imag(x)) < _jinc_threshold(T)
        return evalpoly(x^2, (T(1), -T(π)^2/8, T(π)^4/192))
    else
        return _jinc_core(x)
    end
```
would work just as fine.  You can also use the more standard
```julia
@inline function _jinc(x::Union{T,Complex{T}}) where {T<:Union{Float32,Float64}}
    if abs(real(x)) + abs(imag(x)) < _jinc_threshold(T)
        return evalpoly(x^2, (T(1), -T(π)^2/8, T(π)^4/192))
    else
        return _jinc_core(x)
    end
end
```
For the `evalpoly` issue, you need to use `@evalpoly` instead of `evalpoly`.

Also, for `fastabs` I'd do something like
```julia
function fastabs(x)
    @static if isdefined(Base.Math, :fastabs)
        return Base.Math.fastabs(x)
    else
        # copy definition here....
    end
end
```

File: src/bessel.jl [hunk 2]
```diff
@@ -758,3 +758,44 @@ External links: [DLMF](https://dlmf.nist.gov/10.2.6), [Wikipedia](https://en.wik
 See also: [`hankelh2(nu,x)`](@ref SpecialFunctions.hankelh2)
 """
 hankelh2x(nu, z) = besselhx(nu, 2, z)
+
+"""
+    fastabs(x)
+    
+In future just use `fastabs` from Base.Math
+https://github.com/JuliaLang/julia/blob/93fb785831dcfcc442f82fab8746f0244c5274ae/base/special/trig.jl#L1057
+"""
```
Comment (giordano) [inline src/bessel.jl:767]: This is an internal implementation detail, no need to put in a docstring.  I'd remove the docstring and just add a comment

Comment (roflmaostc) [reply src/bessel.jl]: Sure, I removed it

File: src/bessel.jl [hunk 3]
```diff
@@ -758,3 +758,41 @@ External links: [DLMF](https://dlmf.nist.gov/10.2.6), [Wikipedia](https://en.wik
 See also: [`hankelh2(nu,x)`](@ref SpecialFunctions.hankelh2)
 """
 hankelh2x(nu, z) = besselhx(nu, 2, z)
+
+    
+ # In future just use `fastabs` from Base.Math
+ # https://github.com/JuliaLang/julia/blob/93fb785831dcfcc442f82fab8746f0244c5274ae/base/special/trig.jl#L1057
+function fastabs(x)
```
Comment (stevengj) [inline src/bessel.jl:765]: You can do this the other way:
```jl
if isdefined(Base.Math, :fastabs)
    import Base.Math: fastabs
else
    fastabs(x) = abs(x)
    fastabs(x::Complex) = abs(real(x)) + abs(imag(x)) 
end
```

Comment (stevengj) [reply src/bessel.jl:765]: This definition should probably go into SpecialFunctions.jl and be used in a few other places in the code (e.g. https://github.com/JuliaMath/SpecialFunctions.jl/blob/9116108ba52d103a05d47d9af1ca6b959a536c30/src/gamma.jl#L477 and https://github.com/JuliaMath/SpecialFunctions.jl/blob/9116108ba52d103a05d47d9af1ca6b959a536c30/src/expint.jl#L181 and https://github.com/JuliaMath/SpecialFunctions.jl/blob/9116108ba52d103a05d47d9af1ca6b959a536c30/src/expint.jl#L181)

File: src/bessel.jl [hunk 4]
```diff
@@ -758,3 +758,41 @@ External links: [DLMF](https://dlmf.nist.gov/10.2.6), [Wikipedia](https://en.wik
...
+Following convention: ``\\operatorname{jinc}{x} = \\frac{2 \\cdot J_1{\\pi x}}{\\pi x}``.
+Sometimes known as sombrero or besinc function.
+
+External links: [Wikipedia](https://en.wikipedia.org/wiki/Sombrero_function)
+"""
+jinc(x::Number) = iszero(x) ? oftype(x, 1) : _jinc(float(x))
+_jinc(x::Number) = _jinc_core(x)
+
+# below these thresholds we evaluate a fourth order Taylor polynomial
+_jinc_threshold(::Type{Float64}) = 0.001
```
Comment (stevengj) [inline src/bessel.jl:786]: Looks like 0.002 would be sufficient for machine precision?

File: src/bessel.jl [hunk 5]
```diff
@@ -758,3 +758,41 @@ External links: [DLMF](https://dlmf.nist.gov/10.2.6), [Wikipedia](https://en.wik
...
+_jinc_threshold(::Type{Float32}) = 0.01f0
```
Comment (stevengj) [inline src/bessel.jl:787]: Looks like `0.05f0` would be sufficient here?

File: src/bessel.jl [hunk 6]
```diff
@@ -758,3 +758,41 @@ External links: [DLMF](https://dlmf.nist.gov/10.2.6), [Wikipedia](https://en.wik
...
+"""
+    jinc(x)
+
+Bessel function of the first kind divided by `x`.
+Following convention: ``\\operatorname{jinc}{x} = \\frac{2 \\cdot J_1{\\pi x}}{\\pi x}``.
+Sometimes known as sombrero or besinc function.
+
+External links: [Wikipedia](https://en.wikipedia.org/wiki/Sombrero_function)
+"""
+jinc(x::Number) = iszero(x) ? oftype(x, 1) : _jinc(float(x))
```
Comment (stevengj) [inline src/bessel.jl:782]: This is wrong — the oftype conversion should happen *after* `float(x)`.  That is, this test should happen in `_jinc`, not in `jinc`

Comment (stevengj) [reply src/bessel.jl:782]: And you can just use `one(x)` instead of `oftype(x,1)`.

File: test/bessel.jl [hunk 7]
```diff
@@ -271,4 +271,54 @@ end
...
+            # check for symmetry
+            for x = -5.0:1.0:3.0
+                @test jinc(T(10^x)) ≈ jinc(T(- 10^x))
+
+            end
+
+
+            x = T(1e-5)
+            res = T(0.999999999876629944991456407422485505370463)
+            @test jinc(x) ≈ res
```
Comment (stevengj) [inline test/bessel.jl:294]: `≈` by default only checks that half of the significant digits match.  I think we want a more stringent test here.

Instead, I would use the `≅` comparison operation defined here https://github.com/JuliaMath/SpecialFunctions.jl/blob/9116108ba52d103a05d47d9af1ca6b959a536c30/test/runtests.jl#L13

File: test/bessel.jl [hunk 8]
```diff
@@ -271,4 +271,54 @@ end
...
+
+            x = T(2e-2)
+            res = T(0.9995066009475120760121935544369072906921634576)
+            @test jinc(x) ≈ res
+
+            # type stability for 0
+            @test typeof(jinc(T(0))) == T
```
Comment (stevengj) [inline test/bessel.jl:301]: Make sure you test this for non-float types:
```jl
for T in (Int, Rational{Int}, BigInt,Float16,ComplexF16)
    @test @inferred(jinc(one(T))) === jinc(float(one(T)))
    @test jinc(zero(T)) === float(one(T)))
end
```
(`===` checks that the types match.)

For `Float16`, there should be a method that converts to `Float32` and then rounds back:
```jl
_jinc(x::Float16) = Float16(_jinc(Float32(x)))
_jinc(x::ComplexF16) = ComplexF16(_jinc(ComplexF32(x)))
```

File: src/bessel.jl (hunk 6 above)
Comment (roflmaostc) [reply src/bessel.jl:782]: I'm not sure if I understand you correctly, but: `one(x)` should happen after float to preserve type stability? 
Currently it is `typeof(jinc(0)) != typeof(jinc(1))` and you want to have the same type, right?

File: test/bessel.jl (hunk 7 above)
Comment (roflmaostc) [reply test/bessel.jl:294]: Didn't know about that one, but I'll try to fix it.

Comment (stevengj) [reply test/bessel.jl:294]: We'll need to update `≅` so that it uses a different tolerance for single precision.  Something like:
```jl
checktol(err::Float32) = err ≤ 1e-6
checktol(err::Float64) = err ≤ 1e-13
≅(a,b) = checktol(relerrc(a,b))
```

File: test/bessel.jl (hunk 8 above)
Comment (roflmaostc) [reply test/bessel.jl:301]: Ah, didn't know that.
For me that it feels like cheating to switch to higher precision in between? I assumed that we want to calculate the result using `Float16`. 
I'm not familiar with stuff like this...

File: src/bessel.jl (hunk 3 above)
Comment (roflmaostc) [reply src/bessel.jl:765]: For `gamma.jl` it seems fine, I could include that.
But the other two are slightly different, aren't they?

File: test/bessel.jl (hunk 8 above)
Comment (stevengj) [reply test/bessel.jl:301]: No, it is perfectly fine to switch to a higher-precision in between.   The only reason to use lower precision is if it is faster, but `Float16` is actually slower than `Float32` because only the latter is implemented in hardware on typical CPUs.

(`Float16` is mostly used as a storage format, or on GPUs.)

File: src/bessel.jl (hunk 6 above)
Comment (stevengj) [reply src/bessel.jl:782]: Yes.  That is, you should do
```jl
jinc(x::Number) = _jinc(float(x))
_jinc(x::Number) = iszero(x) ? one(x) : _jinc_core(x)
```

File: src/bessel.jl (hunk 3 above)
Comment (stevengj) [reply src/bessel.jl:765]: They are slightly different, but `fastabs` is fine as a replacement in the expint case.

**Comment by roflmaostc** [comment]
Just found out, that www.wolframalpha seems to struggle provide high precision results...
However, Mathematica produces similar results as Julia :D

**Comment by roflmaostc** [comment]
I could track down the problem

Mathematica:
```mathematica
BesselJ[1, 1.8`200]
```

```julia
besselj1(big(1.8))
```

Outputs:
```
0.581516951731165184221547593970460741118046297653496168575818368562848497914668
0.58151695173116518347008479715619083566653627328448241718624528992169
```

**Comment by roflmaostc** [comment]
Maybe not gootdstyle, but the test cases I pushed  fail. I'm not sure what's going on but the problem is probably somewhere in `besselj1`.

Does someone know why Mathematica (and some webpages) show different results and is it expected?

**Comment by stevengj** [comment]
You want 
```jl
julia> besselj1(big"1.8")
0.5815169517311651834700847971561908356665362732844824171862452899216979041922363
```
The problem is that `big(1.8)` is not the number you are thinking of:
```jl
julia> big(1.8)
1.8000000000000000444089209850062616169452667236328125
```
That is because `1.8` is a `Float64` literal, but 1.8 is not exactly representable as a binary floating-point number so it is rounded to something that differs from 1.8 in the 17th decimal place. In other words, `big(1.8)` means *round 1.8 to the nearest Float64 and *then* convert to BigFloat*.

**Comment by roflmaostc** [comment]
Sorry to bother. It fixed many cases, but not all.
I tried to make sure that it's parsed correctly
```julia
julia> res2 = big"1.971105539196620210542904558564976891448851717386016064466205835417372529649732623778146798009632930245313240967710133352055225553347549825667896762467087506209685330050397030027682762563950964962993e11" - big"5.89917589196246805438832769814196828580119495400655210987269285701113079909408122244782544425726443894463237062342568390611353700647229983489787030738912900107228049806443988000825445873073676204e6"*1im
1.971105539196620210542904558564976891448851717386016064466205835417372529649727e+11 - 5.899175891962468054388327698141968285801194954006552109872692857011130799094111e+06im

julia> Complex{Float64}(res2)
1.9711055391966202e11 - 5.899175891962468e6im

julia> x2 = big"0.00001" + big"10"*1im
9.99999999999999999999999999999999999999999999999999999999999999999999999999994e-06 + 10.0im

julia> jinc(Complex{Float64}(x2))
1.9711055391966183e11 - 5.899175891974533e6im
```

However, the result still mismatches. Any further idea?

**Comment by stevengj** [comment]
> However, the result still mismatches. Any further idea?

`besselj1(Complex{Float64}(x2))` matches the Mathematica result with relative error `1.81e-16`, i.e. nearly to machine precision as measured in the complex plane.  The relative error in the real part by itself is `6.41e-12` and the relative error in the (much larger) imaginary part by itself is `1.70e-16`.

Because the error in the real part by itself is `> 1e-13`, it will fail the `≅` equality test.  But that's not a big cause for concern — `≅` equality is sometimes too stringent.  It's quite common for special functions to be accurate only in the sense of `abs(z-z₀)/abs(z₀)` being small, not for the real and imaginary parts individually, and `≅` equality sometimes fails when the one of the two parts is much smaller than the other (i.e. close to the real or imaginary axes).  Apparently the Bessel functions that we are using fall into this category.

**Comment by roflmaostc** [comment]
Ok, got that.

I relaxed \cong for complex number to check for abs accuracy if real and imag accuracy fails.

**Comment by stevengj** [comment]
> I relaxed \cong for complex number to check for abs accuracy if real and imag accuracy fails.

No, please don't do that!  Please revert this change.

In cases where the stronger accuracy condition holds, we want to continue to ensure it holds!

Only change the tests for the cases that don't currently support this level of accuracy.  In that case, you don't need a new operator at all — you can just use `≈` with an `rtol` argument, e.g.
```jl
@test z ≈ z₀ rtol=1e-13
```

File: src/bessel.jl [hunk 9]
```diff
@@ -758,3 +758,36 @@ External links: [DLMF](https://dlmf.nist.gov/10.2.6), [Wikipedia](https://en.wik
...
+_jinc(x::ComplexF16) = ComplexF16(_jinc(ComplexF32(x)))
+
+# below these thresholds we evaluate a fourth order Taylor polynomial
+_jinc_threshold(::Type{Float64}) = 0.002
+_jinc_threshold(::Type{Float32}) = 0.05f0
+
+
+ # for small arguments, a Taylor series is faster
+@inline function _jinc(x::Union{T,Complex{T}}) where {T<:Union{Float32,Float64}}
+    if iszero(x)
```
Comment (stevengj) [inline src/bessel.jl:783]: This special case is not needed, since the `@evalpoly` branch already handles it.

File: test/expint.jl [hunk 10]
```diff
@@ -32,7 +32,7 @@ using Base.MathConstants
...
         for (x, y) in zip(inputs, sympy_outputs)
             if y == 5.7333170913660523e-42 + 4.019319913618439e-39im
-                @test relerrc(expint(x), y) ≤ 5e-12
+                @test relerr(expint(x), y) ≤ 5e-12
```
Comment (stevengj) [inline test/expint.jl:35]: Why is this changed?

File: test/expint.jl [hunk 11]
```diff
@@ -46,7 +46,7 @@ using Base.MathConstants
 
     @testset "En" begin
         # ≅ but worse
-        ⩭(x, y) = relerrc(x, y) ≤ 5e-12
+        ⩭(x, y) = relerr(x, y) ≤ 5e-12
```
Comment (stevengj) [inline test/expint.jl:49]: Why is this changed?  You are making the tests weaker by changing `relerrc` to `relerr`.

File: test/gamma.jl [hunk 12]
```diff
@@ -172,9 +172,9 @@ end
     @test zeta(4.1+0.3im, -3.2+0.1im) ≅ -281.34474134962502296077659347175501181994490498591796647 + 286.55601240093672668066037366170168712249413003222992205im
     @test zeta(4.1+0.3im, 3.2+0.1im) ≅ 0.0121197525131633219465301571139288562254218365173899270675-0.00687228692565614267981577154948499247518236888933925740902im
     @test zeta(4.1, 3.2+0.1im) ≅ 0.0137637451187986846516125754047084829556100290057521276517-0.00152194599531628234517456529686769063828217532350810111482im
-    @test 1e-12 > relerrc(zeta(1.0001, -4.5e2+3.2im), 10003.765660925877260544923069342257387254716966134385170 - 0.31956240712464746491659767831985629577542514145649468090im)
+    @test 1e-12 > relerr(zeta(1.0001, -4.5e2+3.2im), 10003.765660925877260544923069342257387254716966134385170 - 0.31956240712464746491659767831985629577542514145649468090im)
```
Comment (stevengj) [inline test/gamma.jl:175]: As above — please don't weaken unrelated tests.

File: test/runtests.jl [hunk 13]
```diff
@@ -8,9 +8,13 @@ using SpecialFunctions: AmosException, f64
 
 # useful test functions for relative error, which differ from isapprox
 # in that relerrc separately looks at the real and imaginary parts
```
Comment (stevengj) [inline test/runtests.jl:10]: Oh, I see, you merged `relerr` and `relerrc` — that's fine, then.  But please change the comment as well.

File: test/expint.jl (hunk 11 above)
Comment (roflmaostc) [reply test/expint.jl:49]: Actually I didn't want to weaken the tests.
`relerrc` can be expressed with `relerr` using dynamic dispatch and this is what I was trying to do.
The test cases failed because I messed up locally by not using a fresh REPL.
I push a correct version of `relerr` for complex input

**Comment by stevengj** [comment]
Looks good.  Now you just need to add it to the documentation (at the end of the list of Bessel functions) [here](https://github.com/JuliaMath/SpecialFunctions.jl/blob/master/docs/src/functions_list.md) and [here](https://github.com/JuliaMath/SpecialFunctions.jl/blob/master/docs/src/functions_overview.md).
